app/src/main/java/org/mozilla/fenix/components/metrics/EngineTelemetryMiddleware.kt

```diff
@@ -0,0 +1,67 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.components.metrics
+
+import mozilla.components.browser.state.action.BrowserAction
+import mozilla.components.browser.state.action.EngineAction
+import mozilla.components.browser.state.selector.findTabOrCustomTab
+import mozilla.components.browser.state.state.BrowserState
+import mozilla.components.browser.state.state.EngineState
+import mozilla.components.browser.state.state.SessionState
+import mozilla.components.lib.state.Middleware
+import mozilla.components.lib.state.MiddlewareContext
+import mozilla.components.support.base.android.Clock
+import mozilla.components.support.base.log.logger.Logger
+import org.mozilla.fenix.GleanMetrics.Engine as EngineMetrics
+
+private val logger = Logger("EngineTelemetryMiddleware")
+
+/**
+ * A Middleware to collect some engine-specific (GeckoView) telemetry.
+ *
+ * https://github.com/mozilla-mobile/android-components/issues/9366
+ */
+class EngineTelemetryMiddleware : Middleware<BrowserState, BrowserAction> {
+    override fun invoke(
+        context: MiddlewareContext<BrowserState, BrowserAction>,
+        next: (BrowserAction) -> Unit,
+        action: BrowserAction
+    ) {
+        if (action is EngineAction.KillEngineSessionAction) {
+            val tab = context.state.findTabOrCustomTab(action.sessionId)
+            onEngineSessionKilled(context.state, tab)
+        }
+
+        next(action)
+    }
+
+    private fun onEngineSessionKilled(state: BrowserState, tab: SessionState?) {
+        if (tab == null) {
+            logger.debug("Could not find tab for killed engine session")
+            return
+        }
+
+        val isSelected = tab.id == state.selectedTabId
+        val ageNanos = tab.engineState.ageNanos()
+
+        // Increment the counter of killed foreground/background tabs
+        val tabKillLabel = if (isSelected) { "foreground" } else { "background" }
+        EngineMetrics.tabKills[tabKillLabel].add()
+
+        // Record the age of the engine session of the killed foreground/background tab.
+        if (isSelected && ageNanos != null) {
+            EngineMetrics.killForegroundAge.setRawNanos(ageNanos)
+        } else if (ageNanos != null) {
+            EngineMetrics.killBackgroundAge.setRawNanos(ageNanos)
+        }
+    }
+}
+
+@Suppress("MagicNumber")
+private fun EngineState.ageNanos(): Long? {
```

Comment:
Wait this confused me. This is returning ageMillis, right?

Reply:
Oh I see, `elapsedRealtime` actually returns millis, but you want to record nanos. I guess `nowNanos` should be `nowMillis` and `timeStamp` is millis too?

Reply:
Yeah, you are right, I changed this part of the code a couple of times and now the names do not make any sense. :)

Reply:
Ah OK, let's update to `nowMillis` and `timeStampMillis` then? :)